Main.hs

```diff
@@ -4,26 +4,113 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Number Double
+  | Sqrt Expr
+  | Plus Expr Expr
+  | Minus Expr Expr
+  | Prod Expr Expr
+  | Divide Expr Expr
+  | Exp Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Number x) = show x
+  show (Sqrt x) = "sqrt(" ++ show x ++ ")"
+  show (Plus a b) = "(" ++ show a ++ ") + (" ++ show b ++ ")"
+  show (Minus a b) = "(" ++ show a ++ ") - (" ++ show b ++ ")"
+  show (Prod a b) = "(" ++ show a ++ ") * (" ++ show b ++ ")"
+  show (Divide a b) = "(" ++ show a ++ ") / (" ++ show b ++ ")"
+  show (Exp a b) = "(" ++ show a ++ ") ^ (" ++ show b ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) :: Expr -> Expr -> Bool
+  (==) a b = eval a == eval b
 
 data Error
+  = DivisionByZero Expr
+  | RootOfNegative Expr
+  | ZeroToNegativePower Expr
 
 instance Show Error where
-  show = undefined
+  show e = "ERROR! " ++ case e of
+    DivisionByZero expr -> "Cannot divide by zero in \"" ++ show expr ++ "\""
+    RootOfNegative expr -> "Cannot take a root of a negative number in \"" ++ show expr ++ "\""
+    ZeroToNegativePower expr -> "Cannot raise 0 to a negative power in \"" ++ show expr ++ "\""
 
 instance Eq Error where
-  (==) = undefined
+  (==) (DivisionByZero _) (DivisionByZero _) = True
+  (==) (RootOfNegative _) (RootOfNegative _) = True
+  (==) (ZeroToNegativePower _) (ZeroToNegativePower _) = True
+  (==) _ _ = False
+
+evalBinaryOperands :: Expr -> Expr -> (Double -> Double -> Double)-> Either Error Double
+evalBinaryOperands a b op =
+  case eval a of
+    Left err -> Left err
+    Right eval_a ->
+      case eval b of
+        Left err -> Left err
+        Right eval_b -> Right (op eval_a eval_b)
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Number x)= Right x
+eval expr@(Sqrt x) =
+  case eval x of
+  Left err -> Left err
+  Right eval_x
+      | eval_x >= 0 -> Right (sqrt eval_x)
+      | otherwise -> Left (RootOfNegative expr)
+eval (Plus a b) =
+  case evalBinaryOperands a b (+) of 
+    Left err -> Left err
+    Right res -> Right res
+eval (Minus a b) =
+  case evalBinaryOperands a b (-) of 
+    Left err -> Left err
+    Right res -> Right res
+eval (Prod a b) =
+  case evalBinaryOperands a b (*) of 
+    Left err -> Left err
+    Right res -> Right res
+eval expr@(Divide a b) =
+  case evalBinaryOperands a b (/) of 
+    Left err -> Left err
+    Right res
+      | res == 1 / 0 -> Left (DivisionByZero expr)
+      | otherwise -> Right res
+eval expr@(Exp a b) =
+  case evalBinaryOperands a b (**) of 
+    Left err -> Left err
+    Right res
+      | isNaN res -> Left (RootOfNegative expr)
+      | res == 1 / 0 -> Left (ZeroToNegativePower expr)
```

Comment:
There is a better way to check if the result is an infinity. Hoogle the documentation, you'll find it in the vicinity of `isNaN`.